src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EndpointInfoSourceTests.cs

```diff
@@ -254,70 +256,89 @@ private static void VerifyConnection(AppRunner runner, IEndpointInfo endpointInf
+                await _completionEntriesSemaphore.WaitAsync(token);
+                try
                 {
-                    _addedEndpointInfoSources.Add(new (runner, addedEndpointInfoSource));
+                    _completionEntries.Add(entry);
                     _outputHelper.WriteLine($"[Wait] Register App{runner.AppId}");
                 }
+                finally
+                {
+                    _completionEntriesSemaphore.Release();
+                }
 
                 _outputHelper.WriteLine($"[Wait] Wait for App{runner.AppId} notification");
-                timeoutCancellation.CancelAfter(timeout);
-                IEndpointInfo endpointInfo = await addedEndpointInfoSource.Task;
+                IEndpointInfo endpointInfo = await entry.CompletionSource.Task;
                 _outputHelper.WriteLine($"[Wait] Received App{runner.AppId} notification");
 
                 return endpointInfo;
             }
 
-            public Task OnBeforeResumeAsync(IEndpointInfo endpointInfo, CancellationToken cancellationToken)
+            public Task OnBeforeResumeAsync(IEndpointInfo endpointInfo, CancellationToken token)
             {
                 return Task.CompletedTask;
             }
 
-            public void OnAddedEndpointInfo(IEndpointInfo info)
+            public async Task OnAddedEndpointInfoAsync(IEndpointInfo info, CancellationToken token)
             {
                 _outputHelper.WriteLine($"[Source] Added: {ToOutputString(info)}");
-                
-                lock (_addedEndpointInfoSources)
+
+                await _completionEntriesSemaphore.WaitAsync(token);
+                try
                 {
                     _outputHelper.WriteLine($"[Source] Start notifications for process {info.ProcessId}");
 
-                    foreach (var sourceTuple in _addedEndpointInfoSources.ToList())
+                    // Create a mapping of the process ID tasks to the completion entries
+                    IDictionary<Task<int>, CompletionEntry> map = new Dictionary<Task<int>, CompletionEntry>(_completionEntries.Count);
+                    foreach (CompletionEntry entry in _completionEntries)
+                    {
+                        map.Add(entry.Runner.ProcessIdTask.WithCancellation(token), entry);
+                    }
+
+                    while (map.Count > 0)
                     {
-                        AppRunner runner = sourceTuple.Runner;
-                        _outputHelper.WriteLine($"[Source] Checking App{runner.AppId}");
-                        try
+                        // Wait for any of the process ID tasks to complete.
+                        Task<int> completedTask = await Task.WhenAny(map.Keys);
+
+                        map.Remove(completedTask, out CompletionEntry entry);
+
+                        _outputHelper.WriteLine($"[Source] Checking App{entry.Runner.AppId}");
+
+                        if (completedTask.IsCompletedSuccessfully)
```

Comment:
Should we be logging failures if the task does not complete successfully?

Reply:
No, there can still be other processes that were discovered by the ServerEndpointInfoSource that haven't been processed by a callback yet.